### PR TITLE
Fix missing background on upgrade

### LINKED ISSUE #34216: [Bug]: Background chosen in NC 24 is not migrated to NC 25 upon upgrade
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] Nextcloud Server **is** running on 64bit capable CPU, PHP and OS.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

A custom background set in dashboard of NC 24 is not being migrated to NC 25 upon upgrade.

### Steps to reproduce

1. Have NC 24
2. Select a custom, non-default background in the dashboard
3. Upgrade to NC 25 RC

### Expected behavior

- The old background is taken over 

### Installation method

_No response_

### Operating system

_No response_

### PHP engine version

_No response_

### Web server

_No response_

### Database engine version

_No response_

### Is this bug present after an update or on a fresh install?

_No response_

### Are you using the Nextcloud Server Encryption module?

_No response_

### What user-backends are you using?

- [ ] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

_No response_

### List of activated Apps

```shell
- activity: 2.17.0
  - admin_audit: 1.15.0
  - announcementcenter: 6.4.0-beta.1
  - calendar: 4.0.0-alpha.1
  - calendar_resource_management: 0.3.1
  - circles: 25.0.0-alpha.0
  - cloud_federation_api: 1.8.0
  - collectives: 2.0.0-beta1
  - comments: 1.15.0
  - contacts: 5.0.0-alpha.3
  - contactsinteraction: 1.6.0
  - dashboard: 7.5.0
  - dav: 1.24.0
  - deck: 1.8.0-beta.1
  - droneci_fast_lane: 1.0.2
  - end_to_end_encryption: 1.11.0-beta.1
  - external: 4.1.0
  - federatedfilesharing: 1.15.0
  - federation: 1.15.0
  - files: 1.20.0
  - files_lock: 24.0.1
  - files_pdfviewer: 2.6.0
  - files_rightclick: 1.4.0
  - files_sharing: 1.17.0
  - files_trashbin: 1.15.0
  - files_versions: 1.18.0
  - files_videoplayer: 1.14.0
  - firstrunwizard: 2.14.0
  - forms: 3.0.0-beta.1
  - groupfolders: 13.0.0-beta.1
  - guests: 2.3.0
  - integration_discourse: 1.0.4
  - integration_github: 1.0.5
  - integration_mastodon: 1.0.3
  - integration_reddit: 1.0.5
  - integration_twitter: 1.0.3
  - integration_zammad: 1.0.4
  - logreader: 2.10.0
  - lookup_server_connector: 1.13.0
  - mail: 2.0.0-beta.7
  - nextcloud_announcements: 1.14.0
  - notes: 4.5.1
  - notifications: 2.13.1
  - notify_push: 0.4.0
  - oauth2: 1.13.0
  - password_policy: 1.15.0
  - photos: 1.7.1
  - privacy: 1.9.0
  - provisioning_api: 1.15.0
  - related_resources: 1.0.0
  - richdocuments: 7.0.0-beta.1
  - settings: 1.7.0
  - sharebymail: 1.15.0
  - spreed: 15.0.0-rc.1
  - support: 1.8.0
  - survey_client: 1.13.0
  - suspicious_login: 4.3.0
  - systemtags: 1.15.0
  - tasks: 0.14.4
  - text: 3.6.0
  - theming: 2.0.0
  - twofactor_backupcodes: 1.14.0
  - twofactor_nextcloud_notification: 3.5.0-beta.1
  - twofactor_totp: 7.0.0
  - twofactor_webauthn: 1.0.0-beta.2
  - updatenotification: 1.15.0
  - user_ldap: 1.15.0
  - user_status: 1.5.0
  - viewer: 1.9.0
  - weather_status: 1.5.0
  - workflowengine: 2.7.0
  - zammad_organisation_management: 1.1.3
```

### Nextcloud Signing status

_No response_

### Nextcloud Logs

_No response_

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by Pytal**
Fix ~~https://github.com/nextcloud/server/issues/34458~~ https://github.com/nextcloud/server/issues/34216

Background settings previously were set in the `oc_preferences` table with the appid `"dashboard"`, after https://github.com/nextcloud/server/pull/33733 the settings were queried for with the appid `"theming"`

This migration aims to fix the background loss on upgrade by updating the appid to `"theming"`

### CODE REVIEW

**Comment by Pytal** [comment]
The migration looks right but doesn't seem to work, any ideas why @blizzz?

File: core/Migrations/Version25000Date20221007010957.php [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+	 * @param array $options
+	 */
+	public function postSchemaChange(IOutput $output, Closure $schemaClosure, array $options): void {
+		$qb = $this->connection->getQueryBuilder();
+
+		$qb->update('preferences')
+			->set('appid', 'theming')
+			->where($qb->expr()->eq('appid', $qb->createNamedParameter('dashboard')))
+			->andWhere($qb->expr()->eq('configkey', $qb->createNamedParameter('background')))
+			->orWhere($qb->expr()->eq('configkey', $qb->createNamedParameter('backgroundVersion')));
```
Comment (nickvergessen) [inline core/Migrations/Version25000Date20221007010957.php:61]:
```suggestion
			->andWhere($qb->expr()->orX(
				$qb->expr()->eq('configkey', $qb->createNamedParameter('background')),
				$qb->expr()->eq('configkey', $qb->createNamedParameter('backgroundVersion'))
			));
```

File: core/Migrations/Version25000Date20221007010957.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+/**
+ * User background settings handling was moved from the
+ * dashboard app to the theming app so we migrate the
+ * respective preference values here
+ *
+ */
+class Version25000Date20221007010957 extends SimpleMigrationStep {
+
+	/** @var IDBConnection */
+	protected $connection;
```
Comment (nickvergessen) [inline core/Migrations/Version25000Date20221007010957.php:43]:
```suggestion
	protected IDBConnection $connection;
```

**Review by nickvergessen — CHANGES_REQUESTED** [review]

File: core/Migrations/Version25000Date20221007010957.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	/**
+	 * @param IOutput $output
+	 * @param Closure $schemaClosure The `\Closure` returns a `ISchemaWrapper`
+	 * @param array $options
+	 */
+	public function postSchemaChange(IOutput $output, Closure $schemaClosure, array $options): void {
+		$qb = $this->connection->getQueryBuilder();
+
+		$qb->update('preferences')
+			->set('appid', 'theming')
```
Comment (nickvergessen) [inline core/Migrations/Version25000Date20221007010957.php:58]:
```suggestion
			->set('appid', $qb->createNamedParameter('theming'))
```

Exception: Database error when running migration 25000Date20221007010957 for app core An exception occurred while executing a query: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'theming' in 'field list'

File: core/Migrations/Version25000Date20221007010957.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+	/**
+	 * @param IOutput $output
+	 * @param Closure $schemaClosure The `\Closure` returns a `ISchemaWrapper`
+	 * @param array $options
+	 */
+	public function postSchemaChange(IOutput $output, Closure $schemaClosure, array $options): void {
+		$qb = $this->connection->getQueryBuilder();
+
+		$qb->update('preferences')
+			->set('appid', $qb->createNamedParameter('theming'))
```
Comment (PVince81) [inline core/Migrations/Version25000Date20221007010957.php:56]: assuming no collision / key violation possible ?
should we add an extra where to guarantee that there is not already one under the "theming" name ?

Comment (Pytal) [reply core/Migrations/Version25000Date20221007010957.php:56]: Migrations should not address states introduced due to running on `master` (not released) versions right @nickvergessen?

As this will be backported to 25 and the switch from dashboard to theming is also 25

Comment (nickvergessen) [reply core/Migrations/Version25000Date20221007010957.php:56]: I guess you have to take care of it, because otherwise the update breaks e.g. on our instance when people configured an background already?

Comment (PVince81) [reply core/Migrations/Version25000Date20221007010957.php:56]: it looks like it would need rewriting the query and/or using a loop with maybe catching exceptions
might be too much of a hassle to cover for a potential edge case that might not even happen ?

Comment (PVince81) [reply core/Migrations/Version25000Date20221007010957.php:56]: the issue would only affect users of RC3+ anyway, so maybe we delete all the "theming" keys before migrating the "dashboard" ones, assuming that no one pre-RC3 have the "theming" key in place anyway.

for those post-RC3 it might be ok to ask them to set the background again ? (they were using a pre-release anyway!)

Comment (nickvergessen) [reply core/Migrations/Version25000Date20221007010957.php:56]: Also fine with deleting, but needs to be done before we can merge this.

**Comment by Pytal** [comment]
Probably just my local instance not working but updated and ready for review

**Review by PVince81 — APPROVED** [review]
:+1: as discussed

**Comment by PVince81** [comment]
@blizzz we need this for 25.0.0RC4

**Review by CarlSchwan — APPROVED** [review]

**Review by come-nc — APPROVED** [review]
I would feel better if it checked theming/background is not set before overwriting it.

**Comment by PVince81** [comment]
> I would feel better if it checked theming/background is not set before overwriting it.

this is what we wanted to avoid as it would require iterating over all users and slow down the migration/update

in normal cases the migration would not rerun

File: core/Migrations/Version25000Date20221007010957.php [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+		$orExpr = $qb->expr()->orX(
+			$qb->expr()->eq('configkey', $qb->createNamedParameter('background')),
+			$qb->expr()->eq('configkey', $qb->createNamedParameter('backgroundVersion')),
+		);
+
+		$qb->delete('preferences')
+			->where($qb->expr()->eq('appid', $qb->createNamedParameter('theming')))
+			->andWhere($orExpr);
+
+		$qb->update('preferences')
```
Comment (blizzz) [inline core/Migrations/Version25000Date20221007010957.php:64]: get a new `QueryBuilder` instant, this will have state from the previous operation.

**Comment by PVince81** [comment]
seems CI died, there isn't even a job for this PR.
I'll trigger it with a rebase

File: core/Migrations/Version25000Date20221007010957.php (hunk 5 above)
Comment (PVince81) [reply core/Migrations/Version25000Date20221007010957.php:64]: will this break ? or feel free to push as Chris is not available until much later

File: core/Migrations/Version25000Date20221007010957.php [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+		$cleanUpQuery->delete('preferences')
+			->where($cleanUpQuery->expr()->eq('appid', $cleanUpQuery->createNamedParameter('theming')))
+			->andWhere($orExpr);
+		$cleanUpQuery->executeStatement();
+
+		$updateQuery = $this->connection->getQueryBuilder();
+		$updateQuery->update('preferences')
+			->set('appid', $updateQuery->createNamedParameter('theming'))
+			->where($updateQuery->expr()->eq('appid', $updateQuery->createNamedParameter('dashboard')))
+			->andWhere($orExpr);
```
Comment (nickvergessen) [inline core/Migrations/Version25000Date20221007010957.php:69]: Can not be reused as the variables are not set on the querybuilder object

Comment (blizzz) [reply core/Migrations/Version25000Date20221007010957.php]: sheesh, oversaw that one :grimacing:  thanks

**Review by nickvergessen — APPROVED** [review]

**Comment by PVince81** [comment]
@blizzz did you retest after the latest changes ? just to confirm

**Comment by blizzz** [comment]
> @blizzz did you retest after the latest changes ? just to confirm

No, had to rush to an appointment.

**Comment by blizzz** [comment]
First off, I do not think it fixes the original issue, #34458, as in the description. On RC2 the settings are already stored with the "theming" appid.

```
sqlite> select * from oc_preferences where configkey in ('background', 'backgroundVersion');
admin|theming|background|custom
admin|theming|backgroundVersion|1
```

This fixes #34216 INSTEAD which is also great, no question. The concern of the other one are updates within the 25 series.

**Comment by PVince81** [comment]
@blizzz yes, we decided on purpose to "bother" people who are on RC3+ for the sake of simplifying/accelerating the migration (iterating over all users to find existing ones seemed like overkill). So only users post RC3 will have to re-set their backgrounds.

**Comment by blizzz** [comment]
@PVince81 that's two different things. There will be an upgrade to 25.0.1 later on. Or is it ruled out, that (this patch not considered) it will not bug people when upgrading 25.0.0 → 25.0.1?

**Comment by PVince81** [comment]
@blizzz since this is a migration, it will only ever run once. So people upgrading from 25.0.0 to 25.0.x will already have it marked as run. I do agree that there's a slight concern in case someone would decided to clear the migrations or rerun this one.
Should we make two migrations out of it ?

There's also the option of ignoring the database conflict and only using "update", but then people post RC2->25.0.0 will get an update error because we can't overwrite already existing values, in case they exist. This is only for users who had a non-migrated dashboard wallpaper set and at the same time manually set a theme wallpaper.
Now wondering if it's even possible to get into this state ?

**Comment by PVince81** [comment]
for the sake of exercise, if we go for "update only" with a documented workaround, the query to cleanup existing dashboard would be as follow (untested):
```sql
delete from oc_preferences where appid="dashboard" where configkey in ('background', 'backgroundVersion') where exists (select 1 from oc_preferences where appid="theming" where configkey in ('background', 'backgroundVersion'))
```

**Comment by blizzz** [comment]
Now i cannot reproduce https://github.com/nextcloud/server/issues/34458 anymore on a clean setup. I've had it on our instance.

**Comment by PVince81** [comment]
I've done a local test as follows:

1. Start with v24.0.6 with three users: alice, bob, charles
2. Login as alice and set a dashboard wallpaper
3. Login as bob and set a dashboard wallpaper
4. Update to v25.0.0 RC2
5. As alice, set a theming wallpaper

duplicates exist for bob:
```
MariaDB [nextcloud]> select * from oc_preferences where appid in ('theming', 'dashboard');
+--------+-----------+-------------------+-------------+
| userid | appid     | configkey         | configvalue |
+--------+-----------+-------------------+-------------+
| alice  | dashboard | background        | custom      |
| alice  | dashboard | backgroundVersion | 1           |
| alice  | dashboard | firstRun          | 0           |
| bob    | dashboard | background        | custom      |
| bob    | dashboard | backgroundVersion | 1           |
| bob    | dashboard | firstRun          | 0           |
| bob    | theming   | background        | custom      |
| bob    | theming   | backgroundVersion | 1           |
+--------+-----------+-------------------+-------------+
```

**Comment by PVince81** [comment]
@blizzz I've reproduced it, see last (edited) comment

**Comment by blizzz** [comment]
> @blizzz I've reproduced it, see last (edited) comment

You reproduced https://github.com/nextcloud/server/issues/34216

**Comment by PVince81** [comment]
and because of those duplicates as in https://github.com/nextcloud/server/pull/34461#issuecomment-1277454538 we cannot simply run the update like in the first version of this PR:
```
MariaDB [nextcloud]> update oc_preferences set appid='dashboard' where configkey in ('background', 'backgroundVersion');
ERROR 1062 (23000): Duplicate entry 'bob-dashboard-background' for key 'PRIMARY'
```

so the proposal was to delete the backgrounds that were set post RC, here by Bob, by deleting the matching "theming" keys.

so that in the end only users (early adopters) will have to manually re-set the background

**Comment by PVince81** [comment]
@blizzz please check oc_preferences for my user on c.nc.com, my background is still black as I haven't re-selected one.
then we can see the DB state

**Comment by blizzz** [comment]
> 

```
> select appid, configkey, configvalue from oc_preferences where configkey in ('background', 'backgroundVersion') and userid = '*******';
+---------+-------------------+-------------+
| appid   | configkey         | configvalue |
+---------+-------------------+-------------+
| theming | background        | custom      |
| theming | backgroundVersion | 5           |
+---------+-------------------+-------------+
```

**Comment by blizzz** [comment]
> @blizzz please check oc_preferences for my user on c.nc.com, my background is still black as I haven't re-selected one. then we can see the DB state

I had selected on in rc2 which vanished in rc3 and was all black for me. Maybe pre-condition is coming from 24.

**Comment by PVince81** [comment]
I checked user "Alice" on my test system which would represent that test case and here the background is reset to default instead of black.

Maybe c.nc.com is missing that cloudy default pic ?

**Comment by blizzz** [comment]
> I checked user "Alice" on my test system which would represent that test case and here the background is reset to default instead of black.
> 
> Maybe c.nc.com is missing that cloudy default pic ?

Nope.

**Comment by PVince81** [comment]
@blizzz or maybe one needs to upgrade through RC1, RC2, etc to reach that state.

also would need to check in which key the path to the custom bkg image is set, maybe that one can be broken as well ?

**Comment by blizzz** [comment]
> @blizzz or maybe one needs to upgrade through RC1, RC2, etc to reach that state.
> 
> also would need to check in which key the path to the custom bkg image is set, maybe that one can be broken as well ?

Maybe RC 1 then at most. I went now from 24.0.6 (dark mode, non-default stock background) via rc2 (became default background, changed to custom picked from files) and the to RC3. It stayed. So if there was something with rc (or the beta), it does not really matter now. Will close https://github.com/nextcloud/server/issues/34458 and do another check with this PR to confirm  https://github.com/nextcloud/server/issues/34216 is fixed.

**Comment by blizzz** [comment]
Now I took 24.0.6, uploaded a file and set this as background. Then upgrade to RC3 + this patch. Now I have a black background, so, does not really work.

**Comment by blizzz** [comment]
Files need to be moved 

```
data/$APPDATA_DIR/dashboard/$USERID/background.jpg → data/$APPDATA_DIR/theming/$USERID/background.jpg
```

**Comment by PVince81** [comment]
adding the latest idea from internal discussion with @blizzz:
- add a fallback in BackgroundService that checks both locations in oc_preferences and also on disk
- have a time-insensitive (nightly) background job that migrates the oc_preferences keys and also the image file

there's some local WIP work towards this idea that will be pushed later

**Comment by PVince81** [comment]
update:

the fallback + background job for migration is only for the background file, not the config.
we still migrate the config in this PR

**Comment by PVince81** [comment]
I've redone this test:

1. Start with v24.0.6 with three users: alice, bob, charles
2. Login as alice and set a dashboard wallpaper from a local file
3. Login as bob and set a dashboard wallpaper from a local file
4. Upgrade to v25.0.0 RC2
5. As alice, set a theming wallpaper
6. As charles, set a theming wallpaper
7. Switch to this branch
8. Upgrade
9. Check all user's wallpaper

```
MariaDB [nextcloud]> select * from oc_preferences where appid in ('theming', 'dashboard');
+---------+-----------+-------------------+-------------+
| userid  | appid     | configkey         | configvalue |
+---------+-----------+-------------------+-------------+
| alice   | dashboard | firstRun          | 0           |
| alice   | theming   | background        | custom      |
| alice   | theming   | backgroundVersion | 2           |
| bob     | dashboard | firstRun          | 0           |
| bob     | theming   | background        | custom      |
| bob     | theming   | backgroundVersion | 1           |
| charles | dashboard | firstRun          | 0           |
+---------+-----------+-------------------+-------------+
```
in appdata:
```
dashboard
├── alice
│   └── background.jpg
└── bob
    └── background.jpg
theming
├── 1
│   ├── favIcon-core
│   ├── favIcon-dashboard
│   └── icon-core-filetypes_text.svg
├── alice
│   └── background.jpg
└── charles
    └── background.jpg

3 directories, 5 files
```

- Alice: wallpaper is visible :white_check_mark: 
- Bob: wallpaper is visible **through fallback**: :white_check_mark: 
- Charles: wallpaper is gone, this is by design as discussed for users who created in RC2 :white_check_mark: 

10. Run background jobs
11. Check tree again

```
dashboard
├── alice
└── bob
theming
├── 1
│   ├── favIcon-core
│   ├── favIcon-dashboard
│   └── icon-core-filetypes_text.svg
├── alice
│   └── background.jpg
├── bob
│   └── background.jpg
└── charles
    └── background.jpg

4 directories, 6 files
```

- Alice: wallpaper is visible :white_check_mark: 
- Bob: wallpaper is visible: :white_check_mark: 

Note: sadly I still need to shift+refresh to get the wallpapers to refresh, separate issue

**tldr; works as expected** :+1:

**Comment by PVince81** [comment]
- [ ]  raise a ticket to make the background job more scalable with a lot of users as we don't want to fetch a lot of users every time the job runs @blizzz
